File: python/tree/populating-next-right-pointers-in-each-node-ii[1].py
# Definition for binary tree with next pointer.
# class TreeLinkNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution(object):
    def connect(self, root):
        if root is None:
            return

        parent = root
        while parent:
            first, last = None, None
            while parent:
                for node in [parent.left, parent.right]:
                    if node is None:
                        continue
                    if first is None:
                        first = node
                        last = node
                        continue
                    last.next = node
                    last = node

                parent = parent.next

            parent = first

File: c/tree/populating-next-right-pointers-in-each-node-ii[1].c
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  struct TreeLinkNode *left, *right, *next;
 * };
 *
 */
void connect(struct TreeLinkNode *root) {
    struct TreeLinkNode *parent;
    struct TreeLinkNode *arr[2];
    struct TreeLinkNode *first, *last;
    int idx;

    parent = root;
    while(parent){
        first = NULL;
        last = NULL;

        while(parent){
            idx = 0;
            if(parent->left)
                arr[idx++] = parent->left;
            if(parent->right)
                arr[idx++] = parent->right;

            for(int i = 0; i < idx; ++i){
                if(first == NULL){
                    first = arr[i];
                    last = arr[i];
                    continue;
                }
                last->next = arr[i];
                last = last->next;
            }

            parent = parent->next;
        }

        parent = first;
    }
}
